clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -740,4 +728,148 @@ public double measure(MetricConfig config, long now) {
+                this.enabled = false;
+            }
+        }
+
+        public void close() {
+            synchronized (AbstractCoordinator.this) {
+                this.closed = true;
+                AbstractCoordinator.this.notify();
+            }
+        }
+
+        private void await() {
+            synchronized (AbstractCoordinator.this) {
+                try {
+                    AbstractCoordinator.this.wait();
+                } catch (InterruptedException e) {
+
+                }
+            }
+        }
+
+        private void awaitMs(long waitTimeMs) {
+            synchronized (AbstractCoordinator.this) {
+                try {
+                    AbstractCoordinator.this.wait(waitTimeMs);
+                } catch (InterruptedException e) {
+
+                }
+            }
+        }
+
+        private boolean hasFailed() {
+            return failed.get() != null;
+        }
+
+        private RuntimeException failureCause() {
+            return failed.get();
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (true) {
+                    synchronized (AbstractCoordinator.this) {
+                        if (closed)
+                            return;
+
+                        if (!enabled) {
+                            await();
+                            continue;
+                        }
+
+                        if (state != MemberState.STABLE) {
+                            // the group is not stable (perhaps because we left the group or because the coordinator
+                            // kicked us out), so disable heartbeats and wait for the main thread to rejoin.
+                            disable();
+                            continue;
+                        }
+
+                        client.pollNoWakeup();
+                        long now = time.milliseconds();
+
+                        if (coordinatorUnknown()) {
+                            if (findCoordinatorFuture != null && !findCoordinatorFuture.isDone()) {
+                                awaitMs(retryBackoffMs);
+                                continue;
+                            }
+
+                            findCoordinatorFuture = sendGroupCoordinatorRequest();
+                            findCoordinatorFuture.addListener(new RequestFutureListener<Void>() {
+                                @Override
+                                public void onSuccess(Void value) {
+                                    heartbeat.resetTimeouts(time.milliseconds());
+                                }
+
+                                @Override
+                                public void onFailure(RuntimeException e) {
+
+                                }
+                            });
+                        } else if (heartbeat.sessionTimeoutExpired(now)) {
+                            // the session timeout has expired without seeing a successful heartbeat, so we should
```

Comment:
Thought was to protect us from an unclean connection close (e.g. if the broker is partitioned or hard crashes).